src/packageurl/contrib/url2purl.py

```diff
@@ -265,14 +280,203 @@ def build_sourceforge_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+github_raw_content_pattern = (
+    r"^https?://raw.githubusercontent.com/"
+    r"(?P<namespace>.+)/(?P<name>.+)/(?P<version>.+)/"
+    r"(?P<subpath>.+)$"
+)
+
+
+@purl_router.route(github_raw_content_pattern)
+def build_github_raw_content_purl(uri):
+    """
+    Take `uri` as input
+    example uri: https://raw.githubusercontent.com/volatilityfoundation/dwarf2json/master/LICENSE.txt
+    and returns a `PackageURL` object
+    """
+    return purl_from_pattern('github', github_raw_content_pattern, uri)
+
+
+@purl_router.route("https?://api.github\\.com/repos/.*")
+def build_github_api_purl(uri):
+    """
+    Take `uri` as input
+    example : https://api.github.com/repos/nexB/scancode-toolkit/commits/40593af0df6c8378d2b180324b97cb439fa11d66
+    or https://api.github.com/repos/nexB/scancode-toolkit/ 
+    and returns a `PackageURL` object
+    """
+    segments = base_parsing(uri)
+
+    if not(len(segments) >= 3):
+        return
+    namespace = segments[1]
+    name = segments[2]
+    version = None
+
+    #https://api.github.com/repos/nexB/scancode-toolkit/
+    if len(segments) == 4 and segments[3] != 'commits':
+        version = segments[3]
+
+    #https://api.github.com/repos/nexB/scancode-toolkit/commits/40593af0df6c8378d2b180324b97cb439fa11d66
+    if len(segments) == 5 and segments[3] == "commits":
+        version = segments[4]
+
+    return PackageURL(
+        type='github', namespace=namespace, name=name, version=version
+    )
+
+
+
+github_codeload_pattern = (
+    r"https?://codeload.github.com/"
+    r"(?P<namespace>.+)/(?P<name>.+)/(zip|tar.gz|tar.bz2|.tgz)/v(?P<version>.+)$"
+)
+
+
+@purl_router.route(github_codeload_pattern)
+def build_github_codeload_purl(uri):
+    """
+    Take `uri` as input
+    example :  https://codeload.github.com/nexB/scancode-toolkit/tar.gz/v3.1.1
+    and returns a `PackageURL` object
+    """
+    return purl_from_pattern('github', github_codeload_pattern, uri)
+
+
+@purl_router.route("https?://github\\.com/.*")
+def build_github_purl(uri):
+    """
+    Take `uri` as input 
+    For example:
+    https://github.com/package-url/packageurl-js/tree/master/test/data or
+    https://github.com/package-url/packageurl-js/tree/master or
+    https://github.com/package-url/packageurl-js or
+    https://github.com/nexB/scancode-toolkit/archive/v3.1.1.zip
+    do base parsing of the uri, and extract all purl data from that uri, 
+    then returns a `PackageURL` object with the extracted purl data 
+    """
+    #https://github.com/nexB/scancode-toolkit/archive/v3.1.1.zip
+    regex = r"https?://github.com/(?P<namespace>.+)/(?P<name>.+)/archive/v(?P<version>.+).(zip|tar.gz|tar.bz2|.tgz)"
+    matches = re.search(regex, uri)
+
+    if matches:
+        return purl_from_pattern('github', regex, uri)
+
+    segments = base_parsing(uri)
+
+    if not segments:
+        return
+    name = segments[1]
+    namespace = segments[0]
```

Comment:
Echoing the discussion about Package URLs during our GSoC status call, I would prefer the `namespace` value come before `name` in any locations in the code. You seem to do it in most other places, so this might be the only instance right now.